jerry-core/parser/js/js-parser-expr.c

```diff
@@ -1250,6 +1280,38 @@ parser_parse_unary_expression (parser_context_t *context_p, /**< context */
       parser_parse_class (context_p, false);
       return;
     }
+    case LEXER_KEYW_SUPER:
```

Comment:
(Note: I think these groupings come from the standard. But ES5.1 and ES2015 give different keywords and future reserved words. Not sure which grouping to use in js-lexer.h now that the engine supports multiple versions of the standard.)

Reply:
Good idea, how about this?

```
  /* ES5.1 reserved words. Part of them are used in ES2015 features.
   *
   * Note:
   *          these keywords must form a group after all other keywords. */
LEXER_KEYW_CLASS,              /**< class, used by ES2015 classes */
LEXER_KEYW_EXTENDS,            /**< extends, used by ES2015 classes  */
LEXER_KEYW_SUPER,              /**< super, used by ES2015 classes  */
```

Reply:
I think we have to consolidate these
- ES5.1: https://www.ecma-international.org/ecma-262/5.1/#sec-7.6.1.2
- ES2015: https://www.ecma-international.org/ecma-262/6.0/#sec-future-reserved-words

I tried to do a classification of the different keyword/reserved word groups and this is what I got:

Group 0: Future reserved words in ES5.1 promoted to keywords in ES2015:
- class, extends, super, const, export, import

Group 1: Future reserved words in both ES5.1 and in ES2015:
- enum

Group 2: Future reserved words new in ES2015:
- await

Group 3: Future strict reserved words in ES5.1 to be handled specially in ES2015 (identifier or keyword, depending on context):
- yield, let, static

Group 4: Future strict reserved words both in ES5.1 and in ES2015:
- implements, private, public, interface, package, protected

Handling of groups 1 and 4 should be no problem, they haven't changed. Handling of group 0 should be no problem either because keywords and future reserved words are named differently in theory but handled the same in practice (AFAIK: their grouping symbol `LEXER_FIRST_FUTURE_RESERVED_WORD` is not used anywhere in the code).

However, I don't know enough about the internals of the parser to tell how to deal with groups 2 and 3 _properly_. Not sure what grouping symbols to use, where to check for them, etc.

Reply:
How about separate this classification to a further PR?

Reply:
If it's only about grouping and documentation, then doing it in a follow-up is fine. If it needs more doing, e.g., introducing some new macros to handle group 3 (most importantly: static) properly, then it must be done here.

Reply:
As I see there is no technical limitation to split the list of keywords to group.
As you mentioned _group 0_ and _4_ are good as they are. _Group 2_ can be handled properly without any difficulties.

_Group 3_ is a bit more interesting. In general cases they must behave as _group 4_ keyword (here comes why there will be no difference between _group 3_ and _group 4_ in the lexer), but in the special cases (like `static` inside a class body) it will be the current parsing context responsibility to handle the specific keyword properly. Here [is](https://github.com/jerryscript-project/jerryscript/blob/master/jerry-core/parser/js/js-lexer.c#L2311) what I mean about my previous statement. This kind of pre-scanning will prevent the tokenizer to [raise parser error](https://github.com/jerryscript-project/jerryscript/blob/master/jerry-core/parser/js/js-lexer.c#L554).

To summarize, your suggested groups are feasible.